#13/feat statistics

HyungcheolSim opened the pull request:
## 🌿 PR타입
- [X] 기능 추가
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
- [ ] 기타 사소한 수정

## ❗️관련 이슈[#]
통계 API 구현(#13) 
### 📑 개요
- 기간, 기간+시간 별 작성된 Feed의 수 또는 Feed들의 조회수 총합, 공유 수 총합, 좋아요 수 총합을 확인할 수 있는 API 구현
- QueryDSL로 구현 후 main을 pull 받아 중복되는 부분을 대부분 제가 구현한 부분에서 제거하고 통합한 뒤 정상 실행 확인했습니다!

- 해당하는 데이터가 없는 시간대는 제외하고 출력하도록 했습니다!

###  🧷 변경사항
<!--변경 내용을 적어주세요 (커밋 번호를 적어주세요)-->
- ad8ed4ab
  QueryDSL 설정

- 61197d99
Controller에서 사용할 Response CommonResponse와 쿼리 결과를 받을 ResponseDTO StatisticsResponse 생성

- 74346406
Repository, Service, Controller 구현을 통한 기능 구현

- 7eb92b06
기존 코드 pull 후 중복되는 코드들 main 기준으로 통일
Dummy Data의 여러 Count 값들을 DB에서 직접 임의로 넣어준 후 API 동작 확인

## 📷 스크린샷
Count 값들 넣어주기
![image](https://github.com/Teemo-Wanted/SNS_Feed_Service/assets/58131372/0cce3ed6-a48d-4642-ba1b-d1a54d992baf)

실행 확인1_ 테스트태그2,date,2023-10-25,2023-10-30,count
![image](https://github.com/Teemo-Wanted/SNS_Feed_Service/assets/58131372/9dc723a1-395a-4ab5-a88f-eb4c0b43b800)
실행 확인2_테스트태그1, hour,2023-10-25,2023-10-30,count
![image](https://github.com/Teemo-Wanted/SNS_Feed_Service/assets/58131372/1eff04c7-8c2d-4c65-8cb9-aec1817a0261)



## 👀 기타
QueryDSL을 거의 사용해보지 않아서 많이 미흡했지만 구현했습니다.
수정할 부분이 상당히 많아 보입니다. 피드백 주시면 감사히 반영하겠습니다!!!!!
팀원 분들 코드 보며 많이 배울 수 있었습니다.
리팩토링과 테스트 코드 추가를 위해 issue close는 하지 않겠습니다.

HyungcheolSim committed:
setting : querydsl 설정

HyungcheolSim committed:
feat: 범용 Response, QueryDSL 결과를 받아올 Response 생성

HyungcheolSim committed:
feat: 통계 API 구현

**hashtag 입력값 없을 때 jwt에서 hashtag 추출하여 기본값으로 사용하는 기능 추가 필요

HyungcheolSim committed:
refactor: resolve merge conflict and delete duplicate codes and integrate

CheorHyeon commented on src/main/java/com/wanted/sns_feed_service/feed/controller/FeedController.java:
```diff
@@ -63,4 +69,20 @@ public ResponseEntity<FeedDetailResponse> findFeed(@PathVariable long id) {
+    @GetMapping("/statistics")
+    public ResponseEntity<CommonResponse> getFeedByHashtag(
+            @RequestParam(value = "hashtag", required = false, defaultValue = "기본값") String hashtag,
+            @RequestParam(value = "type") String type,
```
type은 required = true 명시 설정 부탁드립니다! 명시하지 않으면 default가 false입니다!

HyungcheolSim replied:
아 제가 잘못 알고 있었군요!! 감사합니다!

CheorHyeon commented on src/main/java/com/wanted/sns_feed_service/feed/controller/FeedController.java:
```diff
@@ -63,4 +69,20 @@ public ResponseEntity<FeedDetailResponse> findFeed(@PathVariable long id) {
+            @RequestParam(value = "end", required = false) LocalDate end,
+            @RequestParam(value = "value", required = false, defaultValue = "count") String value
+    ) {
+        return ResponseEntity.ok(new CommonResponse("게시글 제목 검색 성공", HttpStatus.OK.value(), feedService.getFeeds(hashtag, type, start, end, value)));
```
검색 기능 보다는 "통계 자료 조회 성공"이 좋을 것 같습니다!

HyungcheolSim replied:
넵 반영하겠습니다!

CheorHyeon commented on src/main/java/com/wanted/sns_feed_service/feed/repository/FeedRepositoryImpl.java:
```diff
@@ -117,4 +125,26 @@ private BooleanExpression typeFilter(Type type) {
+        //날짜/시간 형식 설정
+        StringTemplate formattedDate = getFormattedDate(feed.createdAt, type);
+        //count 일 시 게시물 개수 일자별 출력, 아닐 때는 view count 등 합계 출력
+        return select(Projections.fields(StatisticsResponse.class, formattedDate.as("date"), value.equals("count") ? feed.count().intValue().as("count") : value.equals("view_count") ? feed.viewCount.sum().as("count") : value.equals("like_count") ? feed.likeCount.sum().as("count") : feed.shareCount.sum().as("count")))
```
`feed.count().intValue().as("count")`  부분에서
longValue()도 있을까요? id 속성이 Long이라서 Long 개수로 맞추는게 좋을 것 같습니다!

HyungcheolSim replied:
`feed.count()` 를 사용하면 long type으로 리턴되고 다른 `feed.viewCount.sum()` 등 일 경우 int로 리턴되는 걸 보고 하나만 수정하면 되는 count()의 리턴타입을 int로 바꿔줬었습니다.

그럼 count()가 아닌 경우의 리턴타입을 long으로 바꿔보겠습니다!

CheorHyeon commented on src/main/java/com/wanted/sns_feed_service/feed/FeedService.java:
```diff
@@ -37,4 +43,22 @@ public Feed findById(long id) {
+            throw new IllegalArgumentException("날짜별 검색 30일 이하만 가능합니다.");
+        if (type.equals("hour") && diff > 86400 * 7)
+            throw new IllegalArgumentException("날짜/시간 별 검색은 7일 이하만 가능합니다.");
+        return feedRepository.getFeedStatistics(hashtag, type, start, end, value);
```
예외로 처리해버리면 메세지 규격이 달라져가지고 통일시키는것도 좋지 않을까 싶어요!
작성하신 응답 코드 CommonResponse 엔티티를 활용하는 것도 좋을 것 같아요!

CheorHyeon commented on src/main/resources/application.yml:
```diff
@@ -18,4 +18,9 @@ logging:
+
+custom:
+  jwt:
+    secretKey: secretKeysecretKeysecretKeysecretKeysecretKeysecretKeysecretKeysecretKeysecretKeysecretKey
```
키는 변경할테니 다음에는 application-secret에 넣어주시면 감사하겠습니다 하하

CheorHyeon reviewed (approved):
고생하셨습니다. required 옵션 등 의견 확인 부탁드립니다!

HyungcheolSim committed:
refactor: 팀원 피드백 반영

1. jwt secretKey application-secret으로 이롣
2. 통계 API Response Message 변경
3. 통계 API Controller Parameter required true 추가
4. 통계 API 예외 발생 시 CommonResponse 응답으로 변경
5. 통계 API QueryDSL에서 count의 리턴타입과 x_count의 리턴타입 int로 통일했던 부분->long으로 통일하도록 변경,DTO에서 받아주는 타입 long으로 변경
